File: 4_unit_tests/python_tests/suspension_elements/test_node.py
from vehicle_model.suspension_model.suspension_elements.primary_elements.node import Node
import numpy as np

from unittest import main, TestCase


class TestNode(TestCase):
    def test_node_init_position(self):
        test_node = Node(position=[1, 1, 1])

        self.assertListEqual(test_node.position, [1, 1, 1])
    
    def test_node_translate_position(self):
        test_node = Node(position=[1, 1, 1])
        test_node.translate(translation=[-1, -1, -1])
        
        self.assertListEqual(test_node.position, [0, 0, 0])

    def test_node_translate_initial_position(self):
        test_node = Node(position=[1, 1, 1])
        test_node.translate(translation=[-1, -1, -1])
        
        self.assertListEqual(test_node.initial_position, [1, 1, 1])

    def test_node_rotate_base_angles(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[1, 1, 0])

        update_node.rotate(origin=origin_node, ang_x=0, ang_y=0, ang_z=np.pi)
        self.assertListEqual([round(float(x)) for x in update_node.position], [-1, -1, 0])

    def test_node_rotate_edge_angles(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[0, 0, 0])

        update_node.rotate(origin=origin_node, ang_x=0, ang_y=0, ang_z=np.pi)
        self.assertListEqual([round(float(x)) for x in update_node.position], [0, 0, 0])
    
    def test_node_rotate_ND_angles(self):
        origin_node = Node(position=[1, 1, 1])
        update_node = Node(position=[0, 0, 0])

        update_node.rotate(origin=origin_node, ang_x=np.pi, ang_y=np.pi/2, ang_z=0)
        self.assertListEqual([round(float(x)) for x in update_node.position], [2, 2, 2])
    
    def test_node_rotate_base_vector(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[1, 1, 0])

        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=np.pi)
        self.assertListEqual([round(float(x)) for x in update_node.position], [-1, -1, 0])

    def test_node_rotate_edge_vector(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[0, 0, 0])

        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=0)
        self.assertListEqual([round(float(x)) for x in update_node.position], [0, 0, 0])
    
    def test_node_rotate_repeated_one(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[0, 1, 0])

        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=np.pi/2)
        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=np.pi/2)
        self.assertListEqual([round(float(x)) for x in update_node.position], [-1, 0, 0])
    
    def test_node_rotate_repeated_two(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[0, 1, 0])

        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=np.pi/2)
        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=-np.pi/2)
        self.assertListEqual([round(float(x)) for x in update_node.position], [1, 0, 0])

    def test_node_rotate_repeated_three(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[0, 1, 0])

        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=np.pi/2)
        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=0)
        self.assertListEqual([round(float(x)) for x in update_node.position], [0, 1, 0])

    def test_node_reset_position(self):
        origin_node = Node(position=[1, 1, 1])
        update_node = Node(position=[0, 0, 0])

        update_node.translate(translation=[-1, -1, -1])
        update_node.rotate(origin=origin_node, ang_x=np.pi, ang_y=np.pi/2, ang_z=0)
        update_node.reset()

        self.assertListEqual(update_node.position, [0, 0, 0])
    
    def test_node_addition(self):
        first_node = Node(position=[1, 1, 1])
        second_node = Node(position=[1, 1, 1])

        node_sum = first_node + second_node

        self.assertListEqual(node_sum.position, [2, 2, 2])

    def test_node_subtraction(self):
        first_node = Node(position=[3, 3, 3])
        second_node = Node(position=[2, 2, 2])

        node_diff = first_node - second_node

        self.assertListEqual(node_diff.position, [1, 1, 1])
    
    def test_node_multiplication(self):
        node = Node(position=[1, 1, 1])

        node_product = node * 2

        self.assertListEqual(node_product.position, [2, 2, 2])
    
    def test_node_division(self):
        node = Node(position=[2, 2, 2])

        node_quotient = node / 2

        self.assertListEqual(node_quotient.position, [1, 1, 1])
    
    def test_node_getitem(self):
        node = Node(position=[1, 2, 3])

        self.assertEqual(node[2], 3)
    
    def test_node_setitem(self):
        node = Node(position=[1, 2, 3])
        node[2] = 1

        self.assertEqual(node[2], 1)
    
    def test_node_child(self):
        node = Node(position=[0, 0, 0])
        node_child = Node(position=[0, 0, 1])

        node.add_child(node=node_child)

        self.assertIn(node_child, node.child_nodes)
        
    def test_node_child_translate(self):
        node = Node(position=[0, 0, 0])
        node_child = Node(position=[0, 0, 1])

        node.add_child(node=node_child)

        node.translate(translation=[0, 0, 1])

        self.assertEqual(node_child[2], 2)
    
    def test_node_child_reset(self):
        node = Node(position=[0, 0, 0])
        node_child = Node(position=[0, 0, 1])

        node.add_child(node=node_child)

        node.translate(translation=[0, 0, 1])
        node.reset()

        self.assertEqual(node_child[2], 1)

    def test_node_child_rotate_angles(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[1, 1, 0])

        node_child = Node(position=[-1, -1, 0])

        update_node.add_child(node=node_child)

        update_node.rotate(origin=origin_node, direction=[0, 0, 1], angle=np.pi)
        self.assertListEqual([round(float(x)) for x in node_child.position], [1, 1, 0])

    def test_node_child_rotate_vector(self):
        origin_node = Node(position=[0, 0, 0])
        update_node = Node(position=[1, 1, 0])

        node_child = Node(position=[-1, -1, 0])

        update_node.add_child(node=node_child)

        update_node.rotate(origin=origin_node, ang_x=0, ang_y=0, ang_z=np.pi)
        self.assertListEqual([round(float(x)) for x in node_child.position], [1, 1, 0])
    
    # def test_plotter_base(self):
    #     test_plotter = Plotter()

    #     test_node_1 = Node(position=[1, 1, 1])
    #     test_node_2 = Node(position=[1, -1, 1])
    #     test_node_3 = Node(position=[-1, 1, 1])
    #     test_node_4 = Node(position=[-1, -1, 1])
    #     test_node_5 = Node(position=[1, 1, -1])
    #     test_node_6 = Node(position=[1, -1, -1])
    #     test_node_7 = Node(position=[-1, 1, -1])
    #     test_node_8 = Node(position=[-1, -1, -1])

    #     nodes = [test_node_1, test_node_2, test_node_3, test_node_4, test_node_5, test_node_6, test_node_7, test_node_8]

    #     for node in nodes:
    #         test_plotter.add_node(node=node, radius=0.25)
    #     # test_plotter.show()
    
    # def test_plotter_transform(self):
    #     test_plotter = Plotter()

    #     test_node_1 = Node(position=[1, 1, 1])
    #     test_node_2 = Node(position=[1, -1, 1])
    #     test_node_3 = Node(position=[-1, 1, 1])
    #     test_node_4 = Node(position=[-1, -1, 1])
    #     test_node_5 = Node(position=[1, 1, -1])
    #     test_node_6 = Node(position=[1, -1, -1])
    #     test_node_7 = Node(position=[-1, 1, -1])
    #     test_node_8 = Node(position=[-1, -1, -1])

    #     nodes = [test_node_1, test_node_2, test_node_3, test_node_4, test_node_5, test_node_6, test_node_7, test_node_8]

    #     for node in nodes:
    #         test_plotter.add_node(node=node, radius=0.25)

    #     for node in nodes:
    #         node.rotate(origin=Node(position=[0, 0, 0]), angle_x=np.pi/4, angle_y=np.pi/4)

    #     for node in nodes:
    #         test_plotter.add_node(node=node, radius=0.25, color="black")
    #     # test_plotter.show()

File: unit_tests/cpp_tests/test_node.cpp
#include "test_node.h"
#include <cmath>
#include <vector>

using namespace blaze;

// Test: Node initialization with position
TEST_F(TestNode, test_node_init_position) {
    Node test_node(StaticVector<double, 3UL>{1, 1, 1});
    StaticVector<double, 3UL> expected = {1, 1, 1};
    expectVectorEqual(test_node.position, expected);
}

// Test: Node translation
TEST_F(TestNode, test_node_translate_position) {
    Node test_node(StaticVector<double, 3UL>{1, 1, 1});
    test_node.translate(StaticVector<double, 3UL>{-1, -1, -1});
    StaticVector<double, 3UL> expected = {0, 0, 0};
    expectVectorEqual(test_node.position, expected);
}

// Test: Initial position is preserved after translation
TEST_F(TestNode, test_node_translate_initial_position) {
    Node test_node(StaticVector<double, 3UL>{1, 1, 1});
    test_node.translate(StaticVector<double, 3UL>{-1, -1, -1});
    StaticVector<double, 3UL> expected = {1, 1, 1};
    expectVectorEqual(test_node.initial_position, expected);
}

// Test: Node rotation with base angles (ang_x, ang_y, ang_z)
TEST_F(TestNode, test_node_rotate_base_angles) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{1, 1, 0});
    
    update_node.rotate(origin_node, false, std::nullopt, std::nullopt, 0.0, 0.0, M_PI);
    StaticVector<double, 3UL> expected = {-1, -1, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation edge case (rotating node at origin)
TEST_F(TestNode, test_node_rotate_edge_angles) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{0, 0, 0});
    
    update_node.rotate(origin_node, false, std::nullopt, std::nullopt, 0.0, 0.0, M_PI);
    StaticVector<double, 3UL> expected = {0, 0, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation with non-default angles
TEST_F(TestNode, test_node_rotate_ND_angles) {
    Node origin_node(StaticVector<double, 3UL>{1, 1, 1});
    Node update_node(StaticVector<double, 3UL>{0, 0, 0});
    
    update_node.rotate(origin_node, false, std::nullopt, std::nullopt, M_PI, M_PI/2, 0.0);
    StaticVector<double, 3UL> expected = {2, 2, 2};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation with base vector (direction and angle)
TEST_F(TestNode, test_node_rotate_base_vector) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{1, 1, 0});
    
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, M_PI);
    StaticVector<double, 3UL> expected = {-1, -1, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation edge case with vector (rotating node at origin)
TEST_F(TestNode, test_node_rotate_edge_vector) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{0, 0, 0});
    
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, 0.0);
    StaticVector<double, 3UL> expected = {0, 0, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation repeated (two rotations)
TEST_F(TestNode, test_node_rotate_repeated_one) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{0, 1, 0});
    
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, M_PI/2);
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, M_PI/2);
    StaticVector<double, 3UL> expected = {-1, 0, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation repeated (rotation and reverse rotation)
TEST_F(TestNode, test_node_rotate_repeated_two) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{0, 1, 0});
    
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, M_PI/2);
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, -M_PI/2);
    StaticVector<double, 3UL> expected = {1, 0, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node rotation repeated (rotation and zero rotation)
TEST_F(TestNode, test_node_rotate_repeated_three) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{0, 1, 0});
    
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, M_PI/2);
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, 0.0);
    StaticVector<double, 3UL> expected = {0, 1, 0};
    expectVectorNear(update_node.position, expected);
}

// Test: Node reset position
TEST_F(TestNode, test_node_reset_position) {
    Node origin_node(StaticVector<double, 3UL>{1, 1, 1});
    Node update_node(StaticVector<double, 3UL>{0, 0, 0});
    
    update_node.translate(StaticVector<double, 3UL>{-1, -1, -1});
    update_node.rotate(origin_node, false, std::nullopt, std::nullopt, M_PI, M_PI/2, 0.0);
    update_node.reset();
    StaticVector<double, 3UL> expected = {0, 0, 0};
    expectVectorEqual(update_node.position, expected);
}

// Test: Node addition
TEST_F(TestNode, test_node_addition) {
    Node first_node(StaticVector<double, 3UL>{1, 1, 1});
    Node second_node(StaticVector<double, 3UL>{1, 1, 1});
    
    Node node_sum = first_node + second_node;
    StaticVector<double, 3UL> expected = {2, 2, 2};
    expectVectorEqual(node_sum.position, expected);
}

// Test: Node subtraction
TEST_F(TestNode, test_node_subtraction) {
    Node first_node(StaticVector<double, 3UL>{3, 3, 3});
    Node second_node(StaticVector<double, 3UL>{2, 2, 2});
    
    Node node_diff = first_node - second_node;
    StaticVector<double, 3UL> expected = {1, 1, 1};
    expectVectorEqual(node_diff.position, expected);
}

// Test: Node multiplication
TEST_F(TestNode, test_node_multiplication) {
    Node node(StaticVector<double, 3UL>{1, 1, 1});
    
    Node node_product = node * 2;
    StaticVector<double, 3UL> expected = {2, 2, 2};
    expectVectorEqual(node_product.position, expected);
}

// Test: Node division
TEST_F(TestNode, test_node_division) {
    Node node(StaticVector<double, 3UL>{2, 2, 2});
    
    Node node_quotient = node / 2;
    StaticVector<double, 3UL> expected = {1, 1, 1};
    expectVectorEqual(node_quotient.position, expected);
}

// Test: Node indexing (getitem)
TEST_F(TestNode, test_node_getitem) {
    Node node(StaticVector<double, 3UL>{1, 2, 3});
    
    EXPECT_EQ(node[2], 3);
}

// Test: Node indexing (setitem)
TEST_F(TestNode, test_node_setitem) {
    Node node(StaticVector<double, 3UL>{1, 2, 3});
    node[2] = 1;
    
    EXPECT_EQ(node[2], 1);
}

// Test: Node child addition
TEST_F(TestNode, test_node_child) {
    Node node(StaticVector<double, 3UL>{0, 0, 0});
    Node node_child(StaticVector<double, 3UL>{0, 0, 1});
    
    node.add_child(&node_child);
    
    // Check that child is in child_nodes (we can't directly access private members,
    // but we can verify behavior through translate)
    EXPECT_EQ(node_child[2], 1);
}

// Test: Node child translation
TEST_F(TestNode, test_node_child_translate) {
    Node node(StaticVector<double, 3UL>{0, 0, 0});
    Node node_child(StaticVector<double, 3UL>{0, 0, 1});
    
    node.add_child(&node_child);
    node.translate(StaticVector<double, 3UL>{0, 0, 1});
    
    EXPECT_NEAR(node_child[2], 2.0, 1e-6);
}

// Test: Node child reset
TEST_F(TestNode, test_node_child_reset) {
    Node node(StaticVector<double, 3UL>{0, 0, 0});
    Node node_child(StaticVector<double, 3UL>{0, 0, 1});
    
    node.add_child(&node_child);
    node.translate(StaticVector<double, 3UL>{0, 0, 1});
    node.reset();
    
    EXPECT_NEAR(node_child[2], 1.0, 1e-6);
}

// Test: Node child rotation with angles
TEST_F(TestNode, test_node_child_rotate_angles) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{1, 1, 0});
    Node node_child(StaticVector<double, 3UL>{-1, -1, 0});
    
    update_node.add_child(&node_child);
    update_node.rotate(origin_node, false, StaticVector<double, 3UL>{0, 0, 1}, M_PI);
    StaticVector<double, 3UL> expected = {1, 1, 0};
    expectVectorNear(node_child.position, expected);
}

// Test: Node child rotation with vector
TEST_F(TestNode, test_node_child_rotate_vector) {
    Node origin_node(StaticVector<double, 3UL>{0, 0, 0});
    Node update_node(StaticVector<double, 3UL>{1, 1, 0});
    Node node_child(StaticVector<double, 3UL>{-1, -1, 0});
    
    update_node.add_child(&node_child);
    update_node.rotate(origin_node, false, std::nullopt, std::nullopt, 0.0, 0.0, M_PI);
    StaticVector<double, 3UL> expected = {1, 1, 0};
    expectVectorNear(node_child.position, expected);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
